lib/dao.js

```diff
@@ -624,7 +624,184 @@ DataAccessObject.upsert = function(data, options, cb) {
   }
   return cb.promise;
 };
+/**
+ * Performs upsert based on the instance that matches the where clause
+ . This is a PUT operation and based on
+ * non-primary key. The filter takes one key value pair and updates single matching instance.*
+ */
+DataAccessObject.patchOrCreateWithWhere =
+  DataAccessObject.upsertWithWhere = function(where, data, options, cb) {
+    var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+    if (connectionPromise) { return connectionPromise; }
+    assert(arguments.length >= 1, 'At least one argument is required');
+    if (data === undefined && options === undefined && cb === undefined && arguments.length === 1) {
```

Comment:
IIUC, this branch is handling `upsertWithWhere(data)`. IMO, that's an valid usage of `upsertWithWhere` and should be rejected.